File: bcs-app/backend/apis/resources/namespace.py
# -*- coding: utf-8 -*-
#
# Tencent is pleased to support the open source community by making 蓝鲸智云PaaS平台社区版 (BlueKing PaaS Community Edition) available.
# Copyright (C) 2017-2019 THL A29 Limited, a Tencent company. All rights reserved.
# Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://opensource.org/licenses/MIT
#
# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
# an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
# specific language governing permissions and limitations under the License.
#
from rest_framework.response import Response

from backend.accounts import bcs_perm
from backend.apis.resources.serializers import CreateNamespaceParamsSLZ
from backend.apis.views import NoAccessTokenBaseAPIViewSet
from backend.apps.variable.models import NameSpaceVariable
from backend.resources.namespace import utils as ns_utils
from backend.resources.namespace.constants import K8S_PLAT_NAMESPACE, K8S_SYS_NAMESPACE
from backend.resources.project.constants import ProjectKind
from backend.utils.error_codes import error_codes


class NamespaceViewSet(NoAccessTokenBaseAPIViewSet):
    def list_by_cluster_id(self, request, project_id_or_code, cluster_id):
        namespaces = ns_utils.get_namespaces_by_cluster_id(
            request.user.token.access_token, request.project.project_id, cluster_id
        )
        return Response(namespaces)

    def create_mesos_namespace(self, access_token, username, project_id, cluster_id, ns_name):
        """创建mesos命名空间
        注意: mesos中namespace只是一个概念，不是一个资源；因此，不需要在mesos集群创建
        """
        namespace = ns_utils.create_cc_namespace(access_token, project_id, cluster_id, ns_name, username)
        # TODO: 现阶段不向权限中心注入
        return namespace

    def create_k8s_namespace(self, access_token, username, project_id, cluster_id, ns_name):
        raise error_codes.NotOpen()

    def create_namespace(self, request, project_id_or_code, cluster_id):
        project_id = request.project.project_id
        slz = CreateNamespaceParamsSLZ(data=request.data, context={"project_id": project_id})
        slz.is_valid(raise_exception=True)
        data = slz.data

        access_token = request.user.token.access_token
        username = request.user.username
        project_id = request.project.project_id

        project_kind_name = ProjectKind.get_choice_label(request.project.kind)
        namespace = getattr(self, f"create_{project_kind_name.lower()}_namespace")(
            access_token, username, project_id, cluster_id, data["name"]
        )
        # 创建命名空间下的变量值
        ns_id = namespace["id"]
        NameSpaceVariable.batch_save(ns_id, data["variables"])
        namespace["variables"] = data["variables"]

        return Response(namespace)

    def sync_namespaces(self, request, project_id_or_code, cluster_id):
        """同步集群下命名空间
        NOTE: 先仅处理k8s类型
        """
        project_id = request.project.project_id
        # 统一: 通过cc获取的数据，添加cc限制，区别于直接通过线上直接获取的命名空间
        cc_namespaces = ns_utils.get_namespaces_by_cluster_id(request.user.token.access_token, project_id, cluster_id)
        # 转换格式，方便其他系统使用
        cc_namespace_name_id = {info["name"]: info["id"] for info in cc_namespaces}
        # 获取线上的命名空间
        access_token = request.user.token.access_token
        namespaces = ns_utils.get_k8s_namespaces(access_token, project_id, cluster_id)
        # NOTE: 忽略k8s系统和平台自身的namespace
        namespace_name_list = [
            info["resourceName"]
            for info in namespaces
            if info["resourceName"] not in K8S_SYS_NAMESPACE and info["resourceName"] not in K8S_PLAT_NAMESPACE
        ]
        if not (cc_namespaces and namespaces):
            return Response()
        # 根据namespace和realtime namespace进行删除或创建
        # 删除命名空间
        delete_ns_name_list = set(cc_namespace_name_id.keys()) - set(namespace_name_list)
        delete_ns_id_list = [cc_namespace_name_id[name] for name in delete_ns_name_list]
        self.delete_cc_ns(request, project_id, cluster_id, delete_ns_id_list)

        # 添加命名空间
        add_ns_name_list = set(namespace_name_list) - set(cc_namespace_name_id.keys())
        self.add_cc_ns(request, project_id, cluster_id, add_ns_name_list)

        return Response()

    def add_cc_ns(self, request, project_id, cluster_id, ns_name_list):
        access_token = request.user.token.access_token
        creator = request.user.token.access_token
        perm = bcs_perm.Namespace(request, project_id, bcs_perm.NO_RES, cluster_id)
        for ns_name in ns_name_list:
            data = ns_utils.create_cc_namespace(access_token, project_id, cluster_id, ns_name, creator)
            perm.register(data["id"], f"{ns_name}({cluster_id})")

    def delete_cc_ns(self, request, project_id, cluster_id, ns_id_list):
        """删除存储在CC中的namespace"""
        for ns_id in ns_id_list:
            perm = bcs_perm.Namespace(request, project_id, ns_id)
            perm.delete()
            ns_utils.delete_cc_namespace(request.user.token.access_token, project_id, cluster_id, ns_id)

File: bcs-cc/storage/models/namespace.go
/*
 * Tencent is pleased to support the open source community by making 蓝鲸智云PaaS平台社区版 (BlueKing PaaS Community Edition) available.
 * Copyright (C) 2017-2019 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */

package models

import (
	"bcs-cc/storage"

	"github.com/tuvistavie/structomap"
)

// Namespace :
//go:generate goqueryset -in ${GOFILE} -out qs_${GOFILE}
// gen:qs
type Namespace struct {
	Model
	Name           string `json:"name" gorm:"size:63;unique_index"`
	Creator        string `json:"creator" gorm:"size:31"`
	Description    string `json:"description" sql:"size:127"`
	ProjectID      string `json:"project_id" gorm:"size:32;index"`
	ClusterID      string `json:"cluster_id" gorm:"size:32;index"`
	EnvType        string `json:"env_type" gorm:"size:16"`
	Status         string `json:"status" gorm:"size:16"`
	HasImageSecret bool   `json:"has_image_secret" gorm:"default:false"` // 是否有secret
}

// KubernetesNamespace :
//go:generate goqueryset -in ${GOFILE} -out qs_${GOFILE}
// gen:qs
type KubernetesNamespace struct {
	Model
	Name           string `json:"name" gorm:"size:63;unique_index:idx_name_code"`
	Creator        string `json:"creator" gorm:"size:31"`
	Description    string `json:"description" sql:"size:127"`
	ProjectID      string `json:"project_id" gorm:"size:32"`
	ClusterID      string `json:"cluster_id" gorm:"size:32;unique_index:idx_name_code"`
	EnvType        string `json:"env_type" gorm:"size:16"`
	Status         string `json:"status" gorm:"size:16"`
	HasImageSecret bool   `json:"has_image_secret" gorm:"default:false"`
}

// NamespaceFilterParams :
type NamespaceFilterParams struct {
	ProjectID     string
	ClusterID     string
	DesireAllData string
	Limit         int
	Offset        int
}

// K8SNamespaceUpdate :
type K8SNamespaceUpdate struct {
	Namespace        *KubernetesNamespace
	ImageSecretExist bool
}

// MesosNamespaceUpdate :
type MesosNamespaceUpdate struct {
	Namespace        *Namespace
	ImageSecretExist bool
}

// K8SNamespaceList :
func K8SNamespaceList(filter NamespaceFilterParams) (data []map[string]interface{}, count int, err error) {
	nsQuerySet := NewKubernetesNamespaceQuerySet(storage.GetDefaultSession().DB)
	nsQuerySet = nsQuerySet.ProjectIDEq(filter.ProjectID).ClusterIDEqWithEmpty(filter.ClusterID)
	count, err = nsQuerySet.Count()
	if err != nil {
		return nil, 0, err
	}
	if filter.DesireAllData != "1" {
		nsQuerySet = nsQuerySet.Limit(filter.Limit).Offset(filter.Offset)
	}
	var nsList []KubernetesNamespace
	if err := nsQuerySet.All(&nsList); err != nil {
		return nil, 0, err
	}
	serializer := structomap.New().UseSnakeCase().PickAll()
	serializer = serializer.Pick("CreatedAt", "UpdatedAt", "ID").Omit("Model")
	data, err = serializer.TransformArray(nsList)
	if err != nil {
		return nil, 0, err
	}
	return data, count, nil
}

// MesosNamespaceList :
func MesosNamespaceList(filter NamespaceFilterParams) (data []map[string]interface{}, count int, err error) {
	nsQuerySet := NewNamespaceQuerySet(storage.GetDefaultSession().DB)
	nsQuerySet = nsQuerySet.ProjectIDEq(filter.ProjectID).ClusterIDEqWithEmpty(filter.ClusterID)
	count, err = nsQuerySet.Count()
	if err != nil {
		return nil, 0, err
	}
	if filter.DesireAllData != "1" {
		nsQuerySet = nsQuerySet.Limit(filter.Limit).Offset(filter.Offset)
	}
	var nsList []Namespace
	if err := nsQuerySet.All(&nsList); err != nil {
		return nil, 0, err
	}
	serializer := structomap.New().UseSnakeCase().PickAll()
	serializer = serializer.Pick("CreatedAt", "UpdatedAt", "ID").Omit("Model")
	data, err = serializer.TransformArray(nsList)
	if err != nil {
		return nil, 0, err
	}
	return data, count, nil
}

// RetriveRecord : fetch a k8s namespace info
func (namespace *KubernetesNamespace) RetriveRecord() error {
	nsQuerySet := NewKubernetesNamespaceQuerySet(storage.GetDefaultSession().DB)
	nsQuerySet = nsQuerySet.ProjectIDEq(namespace.ProjectID).ClusterIDEqWithEmpty(namespace.ClusterID)
	if err := nsQuerySet.IDEq(namespace.ID).One(namespace); err != nil {
		return err
	}
	return nil
}

// RetriveRecord : fetch a mesos namespace info
func (namespace *Namespace) RetriveRecord() error {
	nsQuerySet := NewNamespaceQuerySet(storage.GetDefaultSession().DB)
	nsQuerySet = nsQuerySet.ProjectIDEq(namespace.ProjectID).ClusterIDEqWithEmpty(namespace.ClusterID)
	if err := nsQuerySet.IDEq(namespace.ID).One(namespace); err != nil {
		return err
	}
	return nil
}

// DeleteRecord : delete a k8s namespace
func (namespace *KubernetesNamespace) DeleteRecord() error {
	if err := namespace.RetriveRecord(); err != nil {
		return err
	}
	db := storage.GetDefaultSession().DB
	// delete record
	if err := namespace.DeleteUnscoped(db); err != nil {
		return err
	}
	return nil
}

// DeleteRecord : delete a mesos namespace
func (namespace *Namespace) DeleteRecord() error {
	if err := namespace.RetriveRecord(); err != nil {
		return err
	}
	db := storage.GetDefaultSession().DB
	// delete record
	if err := namespace.DeleteUnscoped(db); err != nil {
		return err
	}
	return nil
}

// DeleteK8SClusterNamespaces :delete all namespace in k8s cluster
func DeleteK8SClusterNamespaces(projectID string, clusterID string) error {
	nsQuerySet := NewKubernetesNamespaceQuerySet(storage.GetDefaultSession().DB)
	nsQuerySet = nsQuerySet.ProjectIDEq(projectID).ClusterIDEq(clusterID)
	if err := nsQuerySet.DeleteUnscoped(); err != nil {
		return err
	}
	return nil
}

// DeleteMesosClusterNamespaces : delete all namespace in mesos cluster
func DeleteMesosClusterNamespaces(projectID string, clusterID string) error {
	nsQuerySet := NewNamespaceQuerySet(storage.GetDefaultSession().DB)
	nsQuerySet = nsQuerySet.ProjectIDEq(projectID).ClusterIDEq(clusterID)
	if err := nsQuerySet.DeleteUnscoped(); err != nil {
		return err
	}
	return nil
}

// CreateRecord : create k8s namespace record
func (namespace *KubernetesNamespace) CreateRecord() error {
	db := storage.GetDefaultSession().DB
	if err := namespace.Create(db); err != nil {
		return err
	}
	nsQuerySet := NewKubernetesNamespaceQuerySet(db)
	if err := nsQuerySet.NameEq(namespace.Name).One(namespace); err != nil {
		return err
	}
	return nil
}

// CreateRecord : create mesos namespace record
func (namespace *Namespace) CreateRecord() error {
	db := storage.GetDefaultSession().DB
	if err := namespace.Create(db); err != nil {
		return err
	}
	nsQuerySet := NewNamespaceQuerySet(db)
	if err := nsQuerySet.NameEq(namespace.Name).One(namespace); err != nil {
		return err
	}
	return nil
}

// UpdateRecord :
func (updater *K8SNamespaceUpdate) UpdateRecord() error {
	namespace := updater.Namespace
	nsQuerySet := NewKubernetesNamespaceQuerySet(storage.GetDefaultSession().DB)
	nsQuerySet = nsQuerySet.ProjectIDEq(namespace.ProjectID).ClusterIDEq(namespace.ClusterID).IDEq(namespace.ID)
	k8sUpdater := nsQuerySet.GetUpdater()
	k8sUpdater = k8sUpdater.SetNameWithEmpty(
		namespace.Name,
	).SetEnvTypeWithEmpty(
		namespace.EnvType,
	).SetStatusWithEmpty(
		namespace.Status,
	).SetDescription(namespace.Description)
	if updater.ImageSecretExist {
		k8sUpdater = k8sUpdater.SetHasImageSecret(namespace.HasImageSecret)
	}
	if err := k8sUpdater.Update(); err != nil {
		return err
	}
	if err := namespace.RetriveRecord(); err != nil {
		return err
	}
	return nil
}

// UpdateRecord :
func (updater *MesosNamespaceUpdate) UpdateRecord() error {
	namespace := updater.Namespace
	nsQuerySet := NewNamespaceQuerySet(storage.GetDefaultSession().DB)
	nsQuerySet = nsQuerySet.ProjectIDEq(namespace.ProjectID).ClusterIDEq(namespace.ClusterID).IDEq(namespace.ID)
	mesosUpdater := nsQuerySet.GetUpdater()
	mesosUpdater = mesosUpdater.SetNameWithEmpty(
		namespace.Name,
	).SetEnvTypeWithEmpty(
		namespace.EnvType,
	).SetStatusWithEmpty(
		namespace.Status,
	).SetDescription(namespace.Description)
	if updater.ImageSecretExist {
		mesosUpdater = mesosUpdater.SetHasImageSecret(namespace.HasImageSecret)
	}
	if err := mesosUpdater.Update(); err != nil {
		return err
	}
	if err := namespace.RetriveRecord(); err != nil {
		return err
	}
	return nil
}

// NamespaceListByProject : get namespace list by project id list
func NamespaceListByProject(projectIDList []string) (data []map[string]interface{}, err error) {
	db := storage.GetDefaultSession().DB
	// k8s
	k8sQuerySet := NewKubernetesNamespaceQuerySet(db)
	var k8sData []KubernetesNamespace
	if err := k8sQuerySet.ProjectIDInWithoutError(projectIDList...).All(&k8sData); err != nil {
		return nil, err
	}
	k8sSerializer := structomap.New().UseSnakeCase().Pick("ClusterID", "ProjectID", "ID", "Name")
	k8sMapList, err := k8sSerializer.TransformArray(k8sData)
	if err != nil {
		return nil, err
	}
	// mesos
	mesosQuerySet := NewNamespaceQuerySet(db)
	var mesosData []Namespace
	if err := mesosQuerySet.ProjectIDInWithoutError(projectIDList...).All(&mesosData); err != nil {
		return nil, err
	}
	mesosSerializer := structomap.New().UseSnakeCase().Pick("ClusterID", "ProjectID", "ID", "Name")
	mesosMapList, err := mesosSerializer.TransformArray(mesosData)
	if err != nil {
		return nil, err
	}
	// merge k8s and mesos data
	return append(k8sMapList, mesosMapList...), nil
}
